lib/ansible/modules/files/archive.py

```diff
@@ -211,133 +211,152 @@ def main():
     force_archive = params['force_archive']
     globby = False
     changed = False
     state = 'absent'
 
     # Simple or archive file compression (inapplicable with 'zip' since it's always an archive)
     archive = False
-    successes = []
+    b_successes = []
 
     # Fail early
-    if not HAS_LZMA and format == 'xz':
+    if not HAS_LZMA and fmt == 'xz':
         module.fail_json(msg=missing_required_lib("lzma or backports.lzma", reason="when using xz format"),
                          exception=LZMA_IMP_ERR)
         module.fail_json(msg="lzma or backports.lzma is required when using xz format.")
 
     for path in paths:
-        path = os.path.expanduser(os.path.expandvars(path))
+        b_path = os.path.expanduser(
+            os.path.expandvars(
+                to_bytes(path, errors='surrogate_or_strict')
+            )
+        )
 
         # Expand any glob characters. If found, add the expanded glob to the
         # list of expanded_paths, which might be empty.
-        if ('*' in path or '?' in path):
-            expanded_paths = expanded_paths + glob.glob(path)
+        if (b'*' in b_path or b'?' in b_path):
+            b_expanded_paths.extend(glob.glob(b_path))
             globby = True
 
         # If there are no glob characters the path is added to the expanded paths
         # whether the path exists or not
         else:
-            expanded_paths.append(path)
+            b_expanded_paths.append(b_path)
 
     # Only attempt to expand the exclude paths if it exists
     if exclude_paths:
         for exclude_path in exclude_paths:
-            exclude_path = os.path.expanduser(os.path.expandvars(exclude_path))
+            b_exclude_path = os.path.expanduser(
+                os.path.expandvars(
+                    to_bytes(exclude_path, errors='surrogate_or_strict')
+                )
+            )
 
             # Expand any glob characters. If found, add the expanded glob to the
             # list of expanded_paths, which might be empty.
-            if ('*' in exclude_path or '?' in exclude_path):
-                expanded_exclude_paths = expanded_exclude_paths + glob.glob(exclude_path)
+            if (b'*' in b_exclude_path or b'?' in b_exclude_path):
+                b_expanded_exclude_paths.extend(glob.glob(b_exclude_path))
 
                 # If there are no glob character the exclude path is added to the expanded
                 # exclude paths whether the path exists or not.
             else:
-                expanded_exclude_paths.append(exclude_path)
+                b_expanded_exclude_paths.append(b_exclude_path)
 
-    if not expanded_paths:
-        return module.fail_json(path=', '.join(paths), expanded_paths=', '.join(expanded_paths), msg='Error, no source paths were found')
+    if not b_expanded_paths:
+        return module.fail_json(
+            path=', '.join(paths),
+            expanded_paths=', '.join(to_native(p, errors='surrogate_or_strict') for p in b_expanded_paths),
```

Comment:
note, if expanded paths is large, this might be slow.  It's faster to do it like this, if so:
```suggestion
            expanded_paths=to_native(b', '.join(b_expanded_paths), errors='surrogate_or_strict')
```